src/v/raft/consensus.cc

```diff
@@ -2061,34 +2055,22 @@ consensus::do_append_entries(append_entries_request&& r) {
         // the request was delayed/duplicated). In this case we don't want to
         // truncate, otherwise we might lose already committed data.
 
-        struct find_mismatch_consumer {
-            const consensus& parent;
-            model::offset last_log_offset;
-            model::offset last_matched;
-
-            ss::future<ss::stop_iteration>
-            operator()(const model::record_batch& b) {
-                model::offset last_batch_offset
-                  = last_matched
-                    + model::offset(b.header().last_offset_delta + 1);
-                if (
-                  last_batch_offset > last_log_offset
-                  || parent.get_term(last_batch_offset) != b.term()) {
-                    co_return ss::stop_iteration::yes;
-                }
-                last_matched = last_batch_offset;
-                co_return ss::stop_iteration::no;
-            }
+        model::offset last_matched = adjusted_prev_log_index;
+        chunked_vector<model::record_batch> batches;
+        batches.reserve(r.batches().size());
 
-            model::offset end_of_stream() { return last_matched; }
-        };
+        for (auto& b : r.batches()) {
+            model::offset last_batch_offset
+              = last_matched + model::offset(b.header().last_offset_delta + 1);
+            if (
+              last_batch_offset > last_log_offset
+              || get_term(last_batch_offset) != b.term()) {
```

Comment:
It is a bit annoying that we are checking every batch. Can we just find a mismatch point, or alternatively add a check that there is only one mismatch point (i.e. if we are in the else branch then `batches` should be empty)?